### PR TITLE
Added more scope support facets

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
This pull request has these goals:
1. To flesh out the available facets and their definitions
2. To provide one example scope fixture of each facet
3. Update the domain of for each loops name/value/type
4. Update domain of iteration scope for formal arguments

Once the above is done I'm going to do the bulk fixture recorder we talked about to fill in the other languages.

## Checklist

- [x] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [-] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [-] I have not broken the cheatsheet

### CODE REVIEW

File: queries/python.scm [hunk 1]
```diff
@@ -344,14 +343,6 @@
   (#not-parent-type? @_.removal case_clause)
 ) @_.domain
 
-;;!! for name in value:
```
Comment (AndreasArvidsson) [inline queries/python.scm:347]: The scope fixtures showed me that we had two different definitions for this same pattern. Really shows why we need them :)

File: queries/javascript.core.scm [hunk 2]
```diff
@@ -481,24 +480,33 @@
...
-[
-  (call_expression)
-  (new_expression)
-] @functionCall
+(
+  [
+    (call_expression)
+    (new_expression)
+  ] @functionCall @_.domain.start
+  ";"? @_.domain.end
```
Comment (AndreasArvidsson) [inline queries/javascript.core.scm:488]: Semi was missing from domain

File: queries/javascript.scm [hunk 3]
```diff
@@ -36,7 +36,8 @@
   ;;!  ^^^-------
   (field_definition
     property: (_) @name
-  ) @name.domain.start
+    value: (_)? @value
```
Comment (AndreasArvidsson) [inline queries/javascript.scm:39]: Value scope was missing from fields in javascript

File: queries/javascript.jsx.scm [hunk 4]
```diff
@@ -89,7 +89,8 @@
 ;;!! <aaa bbb="ccc" />
 ;;!       ^^^
 (jsx_attribute
-  (property_identifier) @collectionKey
+  (property_identifier) @collectionKey @_.trailing.start.endOf
```
Comment (AndreasArvidsson) [inline queries/javascript.jsx.scm:92]: Was missing trailing delimiter

File: packages/common/src/scopeSupportFacets/scopeSupportFacets.types.ts [hunk 5]
```diff
@@ -66,6 +113,7 @@ export interface ScopeSupportFacetInfo {
 
 export enum ScopeSupportFacetLevel {
   supported,
+  supportedLegacy,
```
Comment (pokey) [inline packages/common/src/scopeSupportFacets/scopeSupportFacets.types.ts:116]: were you planning to do something with this field?

File: packages/common/src/scopeSupportFacets/typescript.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
+/* eslint-disable @typescript-eslint/naming-convention */
+
+import {
+  LanguageScopeSupportFacetMap,
+  ScopeSupportFacetLevel,
+} from "./scopeSupportFacets.types";
+
+const { supported } = ScopeSupportFacetLevel;
+
+export const typescriptScopeSupport: LanguageScopeSupportFacetMap = {
```
Comment (pokey) [inline packages/common/src/scopeSupportFacets/typescript.ts:10]: This feels like a lie to say we only support these 5 things in TS. We should create a way to "import" support from another language, including the tests, the way we do for our `.scm` files. Shall I file an issue for that?

File: packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/functionCall.constructor.scope [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+---
+
+[Content] =
+[Removal] = 0:0-0:14
+0| new Foo.bar(0);
+  >--------------<
+
+[Domain] = 0:0-0:15
+0| new Foo.bar(0);
+  >---------------<
```
Comment (pokey) [inline packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/functionCall.constructor.scope:11]: This looks wrong

File: packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/functionCall.scope [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+---
+
+[Content] =
+[Removal] = 0:0-0:10
+0| foo.bar(0);
+  >----------<
+
+[Domain] = 0:0-0:11
+0| foo.bar(0);
+  >-----------<
```
Comment (pokey) [inline packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/functionCall.scope:11]: and this too

File: packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/functionCallee.constructor.scope [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+---
+
+[Content] =
+[Removal] = 0:0-0:11
+0| new Foo.bar(0);
+  >-----------<
+
+[Domain] = 0:0-0:15
+0| new Foo.bar(0);
+  >---------------<
```
Comment (pokey) [inline packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/functionCallee.constructor.scope:11]: also wrong

File: packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/functionCallee.scope [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+---
+
+[Content] =
+[Removal] = 0:0-0:7
+0| foo.bar(0);
+  >-------<
+
+[Domain] = 0:0-0:11
+0| foo.bar(0);
+  >-----------<
```
Comment (pokey) [inline packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/functionCallee.scope:11]: wrong too

File: packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/name.field.scope [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+  value = 0;
+}
+---
+
+[#1 Content] = 0:6-0:13
+0| class MyClass {
+        >-------<
+1|   value = 0;
+
+2| }
```
Comment (pokey) [inline packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/name.field.scope:11]: This one is another good argument for fine-grained scopes

File: packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/value.field.scope [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+1|   value = 0;
+            >-<
+2| }
+
+
+[Removal] = 1:9-1:11
+0| class MyClass {
+
+1|   value = 0;
+           >--<
```
Comment (pokey) [inline packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/value.field.scope:18]: This looks wrong

File: queries/javascript.core.scm (hunk 2 above)
Comment (pokey) [reply queries/javascript.core.scm:488]: Oh, right this was the bugs I pointed out in my earlier comments 😅. Why would semi be part of the domain? Semi is part of the containing statement; a call is an expression

File: queries/javascript.core.scm [hunk 13]
```diff
@@ -481,24 +480,33 @@
...
-;;!  -----
-(call_expression
-  function: (_) @functionCallee
-) @_.domain
+;;!  ------
+(
+  (call_expression
+    function: (_) @functionCallee
+  ) @_.domain.start
+  ";"? @_.domain.end
```
Comment (pokey) [inline queries/javascript.core.scm:498]: as above

File: queries/javascript.core.scm [hunk 14]
```diff
@@ -481,24 +480,33 @@
...
-;;!  ---------
-(new_expression
-  (arguments) @functionCallee.end.startOf
-) @functionCallee.start.startOf @_.domain
+;;!  ----------
+(
+  (new_expression
+    (arguments) @functionCallee.end.startOf
+  ) @functionCallee.start.startOf @_.domain.start
+  ";"? @_.domain.end
```
Comment (pokey) [inline queries/javascript.core.scm:508]: and here too

**Review by pokey — COMMENTED** [review]
This is actually pretty comprehensive! I added a few more. In a follow-up, we should

- [ ] add more iteration scopes for name / value / type
- [ ] add `named.method`
- [ ] add some more `"statement"` scope facets

File: packages/common/src/scopeSupportFacets/scopeSupportFacetInfos.ts
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/common/src/scopeSupportFacets/scopeSupportFacetInfos.ts:1]: Agreed

File: packages/common/src/scopeSupportFacets/scopeSupportFacetInfos.ts [hunk 15]
```diff
@@ -7,42 +9,274 @@ export const scopeSupportFacetInfos: Record<
...
+  list: {
+    description: "A list/array",
+    scopeType: "list",
+  },
+  map: {
+    description: "A map/dictionary",
+    scopeType: "map",
+  },
+  statement: {
+    description: "A statement",
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/common/src/scopeSupportFacets/scopeSupportFacetInfos.ts:47]: Agreed

File: packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/name.field.scope (hunk 11 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/name.field.scope:11]: Definitely

File: packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/value.field.scope (hunk 12 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/value.field.scope]: updated

File: packages/common/src/scopeSupportFacets/scopeSupportFacetInfos.ts [hunk 16]
```diff
@@ -7,42 +9,275 @@ export const scopeSupportFacetInfos: Record<
...
+  },
+  anonymousFunction: {
+    description:
+      "An anonymous function, eg a lambda function, an arrow function, etc",
+    scopeType: "anonymousFunction",
+  },
+  "anonymousFunction.lambda": {
+    description: "A lambda function",
+    scopeType: "anonymousFunction",
+  },
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/common/src/scopeSupportFacets/scopeSupportFacetInfos.ts:92]: I believe lambda is the correct technical description and exists in many languages. That javascript calls them arrow functions is besides the point. 
Personally I feel that lambdas is something quite different from just an anonymous function. I want facets for every single nuance if possible.

File: packages/common/src/scopeSupportFacets/scopeSupportFacets.types.ts (hunk 5 above)
Comment (AndreasArvidsson) [reply packages/common/src/scopeSupportFacets/scopeSupportFacets.types.ts:116]: Yes it was used an should be used for every language that still has legacy support. Then we can generate documentation to that fact.

File: packages/common/src/scopeSupportFacets/typescript.ts (hunk 6 above)
Comment (AndreasArvidsson) [reply packages/common/src/scopeSupportFacets/typescript.ts:10]: Please do

File: packages/common/src/scopeSupportFacets/scopeSupportFacets.types.ts (hunk 5 above)
Comment (pokey) [reply packages/common/src/scopeSupportFacets/scopeSupportFacets.types.ts:116]: From what I can tell you're not using it anywhere. Prob worth using it at least once if you're going to introduce it? But maybe I just missed it

Comment (AndreasArvidsson) [reply packages/common/src/scopeSupportFacets/scopeSupportFacets.types.ts:116]: It was used for typescript before we migrated all the scopes. It's now used by python as an example.

File: queries/javascript.core.scm (hunk 2 above)
Comment (pokey) [reply queries/javascript.core.scm:488]: I did some testing on this, and here's the behaviour this PR introduces

image

Note how it only extends the domain to include the semicolon for

- the outermost call, and
- only if the entire statement is a function call.

In particular, when it's the value of an assignment, or part of a bigger expression, it doesn't include the semicolon

There is a certain logic to that, but might be surprising? Also might be surprising if you say "previous call" and that doesn't work? I don't have a strong feeling one way or another. I also might consider going a step further and including the semicolon on the content range itself

If we do keep this behaviour, let's capture the semantics from that screenshot in a scope fixture. I tried putting all four of those lines in a scope fixture, and man we really aren't optimized for more than one scope per fixture 😅. I wonder if we should

1. Only keep the empty marker lines if they're inside a scope, or
2. Not bother to show lines outside of the scope; that's kinda what the line numbers are for

cc/ @josharian on both topics mentioned here 😊

Comment (AndreasArvidsson) [reply queries/javascript.core.scm:488]: For now I would be fine with to just reverting the domains so we can actually merge this. Changing the domain of call seems totally separate from this pull request.

Dropping lines outside of the scope is totally be fine for me

I still haven't heard of any way to actually determine token removal. And I I'm not sure if we not losing information if we do that.

File: packages/common/src/scopeSupportFacets/html.ts [hunk 17]
```diff
@@ -1,3 +1,5 @@
+/* eslint-disable @typescript-eslint/naming-convention */
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/common/src/scopeSupportFacets/html.ts:1]: Personally I don't know if I even like naming convention for keys. I think you should be able to use whatever fits your needs.

[the remark this answers is not in the document]
Comment (pokey) [reply packages/common/src/scopeSupportFacets/html.ts:1]: Yeah that's fair. I wonder if we should grab one of the configs from https://typescript-eslint.io/rules/naming-convention/#ignore-properties-that-require-quotes. Eg the third one looks promising; make exceptions if key has certain chars in it, eg `[- .]`

File: queries/javascript.core.scm (hunk 2 above)
Comment (AndreasArvidsson) [reply queries/javascript.core.scm]: Call domain reverted

File: queries/javascript.core.scm (hunk 13 above)
Comment (AndreasArvidsson) [reply queries/javascript.core.scm]: fixed

File: queries/javascript.core.scm (hunk 14 above)
Comment (AndreasArvidsson) [reply queries/javascript.core.scm]: fixed

**Review by pokey — COMMENTED** [review]
Whoops looks like I forgot to submit these comments

File: queries/javascript.core.scm (hunk 2 above)
Comment (josharian) [reply queries/javascript.core.scm]: > I would be supportive of that as a first step. @josharian any objections? If not, I'll file an issue to trackmr

Seems worth trying. If it creates other issues, we can always adjust from there.

File: packages/common/src/scopeSupportFacets/html.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/common/src/scopeSupportFacets/html.ts:1]: Issue created
https://github.com/cursorless-dev/cursorless/issues/2132

File: packages/common/src/scopeSupportFacets/scopeSupportFacetInfos.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/common/src/scopeSupportFacets/scopeSupportFacetInfos.ts]: Lambda facet removed

File: packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/functionCallee.scope (hunk 10 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/functionCallee.scope]: updated

File: packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/functionCallee.constructor.scope (hunk 9 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/functionCallee.constructor.scope]: updated

File: packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/functionCall.scope (hunk 8 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/javascript/functionCall.scope]: updated

File: queries/javascript.core.scm (hunk 2 above)
Comment (pokey) [reply queries/javascript.core.scm]: Filed

- https://github.com/cursorless-dev/cursorless/issues/2139
- https://github.com/cursorless-dev/cursorless/issues/2140

**Review by pokey — APPROVED** [review]
🎉

**Pull request merged**